### PR TITLE
access log: add connection-level logging

### PR DESCRIPTION
**Pull request by kyessenov**
Signed-off-by: Kuat Yessenov <kuat@google.com>

Description: Introducing connection level logging. Access logs are emitted at the listener level (including for no matching filter chains), and can be combined with either HCM or tcp_proxy access logs. The connection stream info is populated with the same data as from tcp_proxy.
Risk Level: low
Testing: unit
Docs Changes: updated
Release Notes: added
Reference: #7917

### CODE REVIEW

File: source/server/connection_handler_impl.h [hunk 1]
```diff
@@ -192,6 +195,8 @@ class ConnectionHandlerImpl : public Network::ConnectionHandler,
     ActiveConnections& active_connections_;
     Network::ConnectionPtr connection_;
     Stats::TimespanPtr conn_length_;
+    Network::ListenerConfig& config_;
+    std::unique_ptr<StreamInfo::StreamInfo> stream_info_;
```
Comment (kyessenov) [inline source/server/connection_handler_impl.h]: I would like to capture missing filter chains as well, for which connections don't exist. Hence, usin this somewhat awkward ownership.

**Comment by kyessenov** [comment]
Please let me know if this API is looking acceptable, before I can proceed with the rest of the change.

File: api/envoy/api/v2/listener.proto [hunk 2]
```diff
@@ -236,4 +237,8 @@ message Listener {
   // This issue was fixed by `tcp: Avoid TCP syncookie rejected by SO_REUSEPORT socket
   // <https://github.com/torvalds/linux/commit/40a1227ea845a37ab197dd1caffb60b047fa36b1>`_.
   bool reuse_port = 21;
+
+  // Configuration for :ref:`access logs <arch_overview_access_logs>`
+  // emitted by the this listener.
```
Comment (mattklein123) [inline api/envoy/api/v2/listener.proto]: s/the this/this

File: include/envoy/event/dispatcher.h [hunk 3]
```diff
@@ -75,11 +76,13 @@ class Dispatcher {
...
    *        connection. Takes ownership of the socket.
    * @param transport_socket supplies a transport socket to be used by the connection.
+   * @param stream_info info object for the server connection
    * @return Network::ConnectionPtr a server connection that is owned by the caller.
    */
   virtual Network::ConnectionPtr
   createServerConnection(Network::ConnectionSocketPtr&& socket,
-                         Network::TransportSocketPtr&& transport_socket) PURE;
+                         Network::TransportSocketPtr&& transport_socket,
+                         StreamInfo::StreamInfo& stream_info) PURE;
```
Comment (mattklein123) [inline include/envoy/event/dispatcher.h]: Given some of the other issues you have opened about async/blocking/etc. logging is this going to end up needing to be a shared_ptr? Not sure if it's worth it to do now or not. nbd

File: include/envoy/stream_info/stream_info.h [hunk 4]
```diff
@@ -137,6 +137,8 @@ struct ResponseCodeDetailValues {
   // indicates that original "success" headers may have been sent downstream
   // despite the subsequent failure.
   const std::string LateUpstreamReset = "upstream_reset_after_response_started";
+  // The downstream connection was rejected because no filter chain matched it.
+  const std::string FilterChainNotFound = "filter_chain_not_found";
```
Comment (mattklein123) [inline include/envoy/stream_info/stream_info.h]: Please do an independent PR on this.

File: source/common/network/connection_impl.h [hunk 5]
```diff
@@ -128,7 +129,7 @@ class ConnectionImpl : public ConnectionImplBase, public TransportSocketCallback
 
   TransportSocketPtr transport_socket_;
   ConnectionSocketPtr socket_;
-  StreamInfo::StreamInfoImpl stream_info_;
+  StreamInfo::StreamInfo& stream_info_;
```
Comment (mattklein123) [inline source/common/network/connection_impl.h]: ?

File: api/envoy/api/v2/listener.proto [hunk 6]
```diff
@@ -236,4 +237,8 @@ message Listener {
   // This issue was fixed by `tcp: Avoid TCP syncookie rejected by SO_REUSEPORT socket
   // <https://github.com/torvalds/linux/commit/40a1227ea845a37ab197dd1caffb60b047fa36b1>`_.
   bool reuse_port = 21;
+
+  // Configuration for :ref:`access logs <arch_overview_access_logs>`
```
Comment (htuch) [inline api/envoy/api/v2/listener.proto]: Can you add some docs with example log messages, etc?

File: api/envoy/api/v2/listener.proto [hunk 7]
```diff
@@ -236,4 +237,8 @@ message Listener {
...
+  // emitted by the this listener.
+  repeated config.filter.accesslog.v2.AccessLog access_log = 22;
```
Comment (ggreenway) [inline api/envoy/api/v2/listener.proto]: Should this be attached to a FilterChain instead of a listener? There is a lot more context about what the traffic is after the FilterChainMatch has been made.

File: source/common/network/connection_impl.h (hunk 5 above)
Comment (kyessenov) [reply source/common/network/connection_impl.h]: TCP connection holds the unique pointer that it shares with this class. I'm fine re-structuring it if you prefer this connection to own the stream info.

File: api/envoy/api/v2/listener.proto (hunk 7 above)
Comment (kyessenov) [reply api/envoy/api/v2/listener.proto]: I'd like to capture a missing filter chain case.

File: include/envoy/stream_info/stream_info.h (hunk 4 above)
Comment (kyessenov) [reply include/envoy/stream_info/stream_info.h]: Ok.

File: api/envoy/api/v2/listener.proto (hunk 6 above)
Comment (kyessenov) [reply api/envoy/api/v2/listener.proto]: Should I do that in general docs or the API? It's pretty much identical to tcp_proxy logging capability.

**Comment by kyessenov** [comment]
Still working on fixing tcp_proxy integration test, but the rest is ready.

**Comment by kyessenov** [comment]
Fixed all tests. PTAL.

File: api/envoy/api/v2/listener.proto (hunk 6 above)
Comment (mattklein123) [reply api/envoy/api/v2/listener.proto]: This needs a release note. Also you can please add some text to https://www.envoyproxy.io/docs/envoy/latest/intro/arch_overview/observability/access_logging that describes this capability (and potentially fleshes out some of the other current uses)?

File: source/common/network/connection_impl.h (hunk 5 above)
Comment (mattklein123) [reply source/common/network/connection_impl.h]: I see. What I would do here is make a `streamInfo()` virtual function that does the correct return based on the context. That is the typical pattern used.

File: source/common/network/filter_manager_impl.cc [hunk 8]
```diff
@@ -87,6 +88,7 @@ FilterStatus FilterManagerImpl::onWrite(ActiveWriteFilter* filter,
 
   std::list<ActiveWriteFilterPtr>::iterator entry;
   if (!filter) {
+    connection_.streamInfo().addBytesSent(buffer_source.getWriteBuffer().buffer.length());
```
Comment (mattklein123) [inline source/common/network/filter_manager_impl.cc]: I think you probably actually want this to be data that is written to the connection, so should this happen after final filter iteration below?

File: source/extensions/quic_listeners/quiche/quic_filter_manager_connection_impl.h [hunk 9]
```diff
@@ -133,6 +130,7 @@ class QuicFilterManagerConnectionImpl : public Network::ConnectionImplBase {
   // stream write. QUICHE doesn't buffer data in connection, all the data is buffered in stream's
   // send buffer.
   EnvoyQuicSimulatedWatermarkBuffer write_buffer_watermark_simulation_;
+  Buffer::OwnedImpl empty_buffer_;
```
Comment (mattklein123) [inline source/extensions/quic_listeners/quiche/quic_filter_manager_connection_impl.h]: nit per a construct on first use singleton for this.

File: source/server/connection_handler_impl.cc [hunk 10]
```diff
@@ -324,21 +324,32 @@ void ConnectionHandlerImpl::ActiveTcpListener::onAcceptWorker(
...
   // Find matching filter chain.
   const auto filter_chain = config_.filterChainManager().findFilterChain(*socket);
   if (filter_chain == nullptr) {
     ENVOY_LOG(debug, "closing connection: no matching filter chain found");
     stats_.no_filter_chain_match_.inc();
+    stream_info->setResponseFlag(StreamInfo::ResponseFlag::NoRouteFound);
+    stream_info->onRequestComplete();
+    for (const auto& access_log : config_.accessLogs()) {
+      access_log->log(nullptr, nullptr, nullptr, *stream_info);
+    }
```
Comment (mattklein123) [inline source/server/connection_handler_impl.cc]: Please consolidate this duplication into a helper function.

File: source/server/connection_handler_impl.cc [hunk 11]
```diff
@@ -434,6 +448,11 @@ ConnectionHandlerImpl::ActiveTcpConnection::ActiveTcpConnection(
 }
 
 ConnectionHandlerImpl::ActiveTcpConnection::~ActiveTcpConnection() {
+  stream_info_->onRequestComplete();
+  for (const auto& access_log : config_.accessLogs()) {
+    access_log->log(nullptr, nullptr, nullptr, *stream_info_);
+  }
```
Comment (mattklein123) [inline source/server/connection_handler_impl.cc]: Can this happen in deferred deletion context? Is this still safe?

File: include/envoy/event/dispatcher.h (hunk 3 above)
Comment (kyessenov) [reply include/envoy/event/dispatcher.h]: Can we postpone adding shared pointer until we need it? I am not sure if it's the right granularity for capturing async logging.

Comment (mattklein123) [reply include/envoy/event/dispatcher.h]: Yup sure.

File: source/common/network/connection_impl.h (hunk 5 above)
Comment (kyessenov) [reply source/common/network/connection_impl.h]: Thanks for explaining. Can you elaborate? I don't know which context the stream info off in ConnectionImpl. Looks like it only has a dispatcher and low-level connection and socket.

File: api/envoy/api/v2/listener.proto (hunk 6 above)
Comment (kyessenov) [reply api/envoy/api/v2/listener.proto]: Done.

File: source/common/network/filter_manager_impl.cc (hunk 8 above)
Comment (kyessenov) [reply source/common/network/filter_manager_impl.cc]: Excellent point, thanks. I think we want "black box" telemetry: bytes read before transformation and bytes sent after transformation.

File: source/common/network/connection_impl.h (hunk 5 above)
Comment (mattklein123) [reply source/common/network/connection_impl.h]: Sorry this is OK. I forgot that there is no ServerConnectionImpl class, just ClientConnectionImpl.

File: source/server/connection_handler_impl.cc (hunk 10 above)
Comment (kyessenov) [reply source/server/connection_handler_impl.cc]: Done

File: source/server/connection_handler_impl.cc (hunk 11 above)
Comment (kyessenov) [reply source/server/connection_handler_impl.cc]: I think it should be fine. Stream info is a unique pointer owned by this object and access logs are  shared pointers.

**Comment by kyessenov** [comment]
Thanks for the review. I'd like to follow-up with smaller changes that I took away to keep this PR manageable:
- add error flags for HCM codec TODOs
- set protocol field for connection stream info and add an access log filter based on it
- add filter chain response code details
- add tests for filter chain and direct response code details

File: source/extensions/quic_listeners/quiche/quic_filter_manager_connection_impl.h [hunk 12]
```diff
@@ -120,6 +117,8 @@ class QuicFilterManagerConnectionImpl : public Network::ConnectionImplBase {
   // Called when aggregated buffered bytes across all the streams declines to low watermark.
   void onSendBufferLowWatermark();
 
+  static Buffer::Instance& emptyBuffer() { MUTABLE_CONSTRUCT_ON_FIRST_USE(Buffer::OwnedImpl); }
```
Comment (mattklein123) [inline source/extensions/quic_listeners/quiche/quic_filter_manager_connection_impl.h]: Sorry I think I led your astray here. I didn't realize this is mutable. If mutable this seems not great so probably revert this? cc @danzh2010

Comment (kyessenov) [reply source/extensions/quic_listeners/quiche/quic_filter_manager_connection_impl.h]: Sure, reverted back.

**Comment by kyessenov** [comment]
Thanks for the review! Can we merge this PR? I'd like to start the follow-ups.
